IllegalStateException in PcmBeagleMappings#addPcmIdOf

An injection Exception caused by a IllegalStateException is thrown when trying to Analyse CocoMe with Beagle.

```
Caused by: java.lang.IllegalStateException
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmBeagleMappings.addPcmIdOf(PcmBeagleMappings.java:192)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.addInternalActionToSet(PcmRepositorySeffExtractor.java:186)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBehaviourAndAddToSet(PcmRepositorySeffExtractor.java:119)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBranchAction(PcmRepositorySeffExtractor.java:163)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBehaviourAndAddToSet(PcmRepositorySeffExtractor.java:124)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBranchAction(PcmRepositorySeffExtractor.java:163)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBehaviourAndAddToSet(PcmRepositorySeffExtractor.java:124)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBranchAction(PcmRepositorySeffExtractor.java:163)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBehaviourAndAddToSet(PcmRepositorySeffExtractor.java:124)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBranchAction(PcmRepositorySeffExtractor.java:163)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositorySeffExtractor.extractBehaviourAndAddToSet(PcmRepositorySeffExtractor.java:124)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositoryExtractor.extractResourceDemandingSEFFImplAndAddContentsToSet(PcmRepositoryExtractor.java:237)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositoryExtractor.extractBasicComponentAndAddContentsToSet(PcmRepositoryExtractor.java:223)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositoryExtractor.scanRepository(PcmRepositoryExtractor.java:207)
at de.uka.ipd.sdq.beagle.core.pcmconnection.PcmRepositoryExtractor.getBlackboardForAllElements(PcmRepositoryExtractor.java:100)
```

> Ja, die IllegalStateException wird geworfen, weil Somox unterschiedliche Seff Elemente auf die gleiche CodeSections mappt (zumindest das was in den EntityNames steht, aus denen ich parse)

> Wenn wir das SourceCode Mapping mit dem Decorator hinbekommen, sollte das Problem gefixt sein. Andernfalls muss es auskommentiert werden oder Somox überarbeitet werden.

--- a/Core/src/main/java/de/uka/ipd/sdq/beagle/core/pcmconnection/PcmRepositorySeffExtractor.java
+++ b/Core/src/main/java/de/uka/ipd/sdq/beagle/core/pcmconnection/PcmRepositorySeffExtractor.java
@@ -43,6 +43,12 @@ public class PcmRepositorySeffExtractor {
 	private static final FailureHandler FAILURE_HANDLER = FailureHandler.getHandler("Beagle FileNotFound Handler");
 
 	/**
+	 * If Somox' generates no more output with multiple identifiers relating to the same
+	 * seffElements, this can be set to true.
+	 */
+	private static final boolean SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS = false;
+	
+	/**
 	 * Temporary storage for all extracted {@link SeffLoop SeffLoops} that should be
 	 * written on the {@link Blackboard}.
 	 */
@@ -93,11 +99,6 @@ public class PcmRepositorySeffExtractor {
 	 */
 	private final PcmCodeSectionGenerator codeSectionGenerator;
 
-	/**
-	 * If Somox' generates no more output with multiple identifiers relating to the same
-	 * seffElements, this can be set to true.
-	 */
-	private final boolean somoxIdentifiersAreUnambiguous = false;
 
 	/**
 	 * Constructor needs access to the real sets (no copy!), manipulating them by adding
@@ -381,7 +382,7 @@ public class PcmRepositorySeffExtractor {
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final ResourceDemandingInternalAction rdia,
 		final String identifier) {
-		if (this.somoxIdentifiersAreUnambiguous) {
+		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {
 			return false;
 		}
 
@@ -402,7 +403,7 @@ public class PcmRepositorySeffExtractor {
 	 *         {@false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final ExternalCallParameter exParam, final String identifier) {
-		if (this.somoxIdentifiersAreUnambiguous) {
+		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {
 			return false;
 		}
 
@@ -424,7 +425,7 @@ public class PcmRepositorySeffExtractor {
 	 *         {@false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final SeffBranch seffBranch, final String identifier) {
-		if (this.somoxIdentifiersAreUnambiguous) {
+		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {
 			return false;
 		}
 
@@ -446,7 +447,7 @@ public class PcmRepositorySeffExtractor {
 	 *         {@false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final SeffLoop seffLoop, final String identifier) {
-		if (this.somoxIdentifiersAreUnambiguous) {
+		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {
 			return false;
 		}
 

--- a/Core/src/main/java/de/uka/ipd/sdq/beagle/core/pcmconnection/PcmRepositorySeffExtractor.java
+++ b/Core/src/main/java/de/uka/ipd/sdq/beagle/core/pcmconnection/PcmRepositorySeffExtractor.java
@@ -376,9 +376,9 @@ public class PcmRepositorySeffExtractor {
 	 *
 	 * @param rdia the seffElement to check
 	 * @param identifier the Identifier to check
-	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
-	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
-	 *         {@false}.
+	 * @return {@code true} if there are no collisions otherwise {@code false}. May be shut down by
+	 *         the variable {@link #SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS} and returning always
+	 *         {@code false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final ResourceDemandingInternalAction rdia,
 		final String identifier) {
@@ -398,9 +398,9 @@ public class PcmRepositorySeffExtractor {
 	 *
 	 * @param exParam the seffElement to check
 	 * @param identifier the Identifier to check
-	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
-	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
-	 *         {@false}.
+	 * @return {@code true} if there are no collisions otherwise {@code false}. May be shut down by
+	 *         the variable {@link #SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS} and returning always
+	 *         {@code false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final ExternalCallParameter exParam, final String identifier) {
 		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {
@@ -420,9 +420,9 @@ public class PcmRepositorySeffExtractor {
 	 *
 	 * @param seffBranch the seffElement to check
 	 * @param identifier the Identifier to check
-	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
-	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
-	 *         {@false}.
+	 * @return {@code true} if there are no collisions otherwise {@code false}. May be shut down by
+	 *         the variable {@link #SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS} and returning always
+	 *         {@code false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final SeffBranch seffBranch, final String identifier) {
 		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {
@@ -442,9 +442,9 @@ public class PcmRepositorySeffExtractor {
 	 *
 	 * @param seffLoop the seffElement to check
 	 * @param identifier the Identifier to check
-	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
-	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
-	 *         {@false}.
+	 * @return {@code true} if there are no collisions otherwise {@code false}. May be shut down by
+	 *         the variable {@link #SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS} and returning always
+	 *         {@code false}.
 	 */
 	private boolean shouldBeIgnoredDueToSomoxCollision(final SeffLoop seffLoop, final String identifier) {
 		if (SOMOX_IDENTIFIERS_ARE_UNAMBIGUOUS) {

--- a/Core/src/main/java/de/uka/ipd/sdq/beagle/core/pcmconnection/PcmRepositorySeffExtractor.java
+++ b/Core/src/main/java/de/uka/ipd/sdq/beagle/core/pcmconnection/PcmRepositorySeffExtractor.java
@@ -94,6 +94,12 @@ public class PcmRepositorySeffExtractor {
 	private final PcmCodeSectionGenerator codeSectionGenerator;
 
 	/**
+	 * If Somox' generates no more output with multiple identifiers relating to the same
+	 * seffElements, this can be set to true.
+	 */
+	private final boolean somoxIdentifiersAreUnambiguous = false;
+
+	/**
 	 * Constructor needs access to the real sets (no copy!), manipulating them by adding
 	 * all extracted SeffElements.
 	 *
@@ -196,9 +202,16 @@ public class PcmRepositorySeffExtractor {
 
 		try {
 			final CodeSection codeSection = this.codeSectionGenerator.getCodeSectionForID(internalAction.getId());
-
 			final ResourceDemandingInternalAction rdiaCpu =
 				new ResourceDemandingInternalAction(ResourceDemandType.RESOURCE_TYPE_CPU, codeSection);
+
+			if (this.shouldBeIgnoredDueToSomoxCollision(rdiaCpu, internalAction.getId())) {
+				return;
+				// The other rdiaTypes do not need to be analysed, because if one of them
+				// would be present in the mapping,
+				// rdiaCPU would be present, too.
+			}
+
 			this.rdiaSet.add(rdiaCpu);
 			this.pcmMapper.addPcmIdOf(rdiaCpu, internalAction.getId());
 
@@ -244,8 +257,12 @@ public class PcmRepositorySeffExtractor {
 
 		try {
 			final CodeSection codeSection = this.codeSectionGenerator.getCodeSectionForID(externalAction.getId());
-
 			final ExternalCallParameter exCallParam = new ExternalCallParameter(codeSection, this.zero);
+
+			if (this.shouldBeIgnoredDueToSomoxCollision(exCallParam, externalAction.getId())) {
+				return;
+			}
+
 			this.externalCallParameterSet.add(exCallParam);
 			this.pcmMapper.addPcmIdOf(exCallParam, externalAction.getId());
 		} catch (final FileNotFoundException exception) {
@@ -283,6 +300,11 @@ public class PcmRepositorySeffExtractor {
 					.getCodeSectionForID(branchTransition.getBranchBehaviour_BranchTransition().getId()));
 			}
 			final SeffBranch seffBranch = new SeffBranch(codeSectionSet);
+			
+			if (this.shouldBeIgnoredDueToSomoxCollision(seffBranch, branchAction.getId())) {
+				return;
+			}
+			
 			this.seffBranchSet.add(seffBranch);
 			this.pcmMapper.addPcmIdOf(seffBranch, branchAction.getId());
 		} catch (final FileNotFoundException exception) {
@@ -324,6 +346,11 @@ public class PcmRepositorySeffExtractor {
 			final CodeSection codeSection =
 				this.codeSectionGenerator.getCodeSectionForID(loopAction.getBodyBehaviour_Loop().getId());
 			final SeffLoop seffLoop = new SeffLoop(codeSection);
+			
+			if (this.shouldBeIgnoredDueToSomoxCollision(seffLoop, loopAction.getId())) {
+				return;
+			}
+			
 			this.seffLoopSet.add(seffLoop);
 			this.pcmMapper.addPcmIdOf(seffLoop, loopAction.getId());
 		} catch (final FileNotFoundException exception) {
@@ -340,6 +367,94 @@ public class PcmRepositorySeffExtractor {
 		 * loopAction.getId()); } } catch (final FileNotFoundException fileNotFoundE) {
 		 * this.handleFailureFor(loopAction, fileNotFoundE); }
 		 */
+	}
+
+	/**
+	 * Testing method to prevent Beagle to fail due to Somox producing ambiguous
+	 * identifiers.
+	 *
+	 * @param rdia the seffElement to check
+	 * @param identifier the Identifier to check
+	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
+	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
+	 *         {@false}.
+	 */
+	private boolean shouldBeIgnoredDueToSomoxCollision(final ResourceDemandingInternalAction rdia,
+		final String identifier) {
+		if (this.somoxIdentifiersAreUnambiguous) {
+			return false;
+		}
+
+		if (this.pcmMapper.hasPcmIdOf(rdia) && !this.pcmMapper.getPcmIdOf(rdia).equals(identifier)) {
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Testing method to prevent Beagle to fail due to Somox producing ambiguous
+	 * identifiers.
+	 *
+	 * @param exParam the seffElement to check
+	 * @param identifier the Identifier to check
+	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
+	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
+	 *         {@false}.
+	 */
+	private boolean shouldBeIgnoredDueToSomoxCollision(final ExternalCallParameter exParam, final String identifier) {
+		if (this.somoxIdentifiersAreUnambiguous) {
+			return false;
+		}
+
+		if (this.pcmMapper.hasPcmIdOf(exParam) && !this.pcmMapper.getPcmIdOf(exParam).equals(identifier)) {
+			return true;
+		}
+
+		return false;
+	}
+	
+	/**
+	 * Testing method to prevent Beagle to fail due to Somox producing ambiguous
+	 * identifiers.
+	 *
+	 * @param seffBranch the seffElement to check
+	 * @param identifier the Identifier to check
+	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
+	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
+	 *         {@false}.
+	 */
+	private boolean shouldBeIgnoredDueToSomoxCollision(final SeffBranch seffBranch, final String identifier) {
+		if (this.somoxIdentifiersAreUnambiguous) {
+			return false;
+		}
+
+		if (this.pcmMapper.hasPcmIdOf(seffBranch) && !this.pcmMapper.getPcmIdOf(seffBranch).equals(identifier)) {
+			return true;
+		}
+
+		return false;
+	}
+	
+	/**
+	 * Testing method to prevent Beagle to fail due to Somox producing ambiguous
+	 * identifiers.
+	 *
+	 * @param seffLoop the seffElement to check
+	 * @param identifier the Identifier to check
+	 * @return {@true} if there are no collisions otherwise {@false}. May be shut down by
+	 *         the variable {@link #somoxIdentifiersWorkCorrectly} and returning always
+	 *         {@false}.
+	 */
+	private boolean shouldBeIgnoredDueToSomoxCollision(final SeffLoop seffLoop, final String identifier) {
+		if (this.somoxIdentifiersAreUnambiguous) {
+			return false;
+		}
+
+		if (this.pcmMapper.hasPcmIdOf(seffLoop) && !this.pcmMapper.getPcmIdOf(seffLoop).equals(identifier)) {
+			return true;
+		}
+
+		return false;
 	}
 
 	/**

